### PR TITLE
Add PPO for cartpole example

### PR DESCRIPTION
**Pull request by iliachry**
This pull request adds a code example of an RL PPO agent solving the CartPole environment.

### CODE REVIEW

**Comment by iliachry** [comment]
@googlebot I signed it!

File: examples/rl/ppo_cartpole.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+[OpenAI Spinning Up Github - PPO](https://github.com/openai/spinningup/blob/master/spinup/algos/tf1/ppo/ppo.py)
+"""
+
+"""
+## Libraries
+
+For this example the following libraries are used:
+
+1. Numpy for n-dimensional arrays
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:53]: NumPy

File: examples/rl/ppo_cartpole.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+2. Tensorflow and Keras for building the deep RL PPO agent
+3. Gym for getting everything we need about the environment
+4. Scipy.signal for calculating the discounted cumulative sums of vectors
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:56]: Use backticks around `scipy.signal`

File: examples/rl/ppo_cartpole.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        # Buffer initialization
+        self.obs_buf = np.zeros((size, obs_dim), dtype=np.float32)
+        self.act_buf = np.zeros(size, dtype=np.float32)
+        self.adv_buf = np.zeros(size, dtype=np.float32)
+        self.rew_buf = np.zeros(size, dtype=np.float32)
+        self.ret_buf = np.zeros(size, dtype=np.float32)
+        self.val_buf = np.zeros(size, dtype=np.float32)
+        self.logp_buf = np.zeros(size, dtype=np.float32)
+        self.gamma, self.lam = gamma, lam
+        self.ptr, self.tr_start_idx = 0, 0
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:87]: Always use fully-spelled out variable names to ensure readability.

File: examples/rl/ppo_cartpole.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    def store(self, obs, act, rew, val, logp):
+        # Append one step of agent-environment interaction
+        self.obs_buf[self.ptr] = obs
+        self.act_buf[self.ptr] = act
+        self.rew_buf[self.ptr] = rew
+        self.val_buf[self.ptr] = val
+        self.logp_buf[self.ptr] = logp
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:95]: For instance, this section will benefit greatly from explicit names.

File: examples/rl/ppo_cartpole.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+"""
+
+# Initialize the environment and get the dimensionality of the
+# observation space and the number of possible actions
+env = gym.make("CartPole-v0")
+obs_dim = env.observation_space.shape[0]
+n_acts = env.action_space.n
+
+# Initialize the buffer
+buf = Buffer(obs_dim, steps_per_epoch)
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:164]: e.g. `buffer`, etc.

File: examples/rl/ppo_cartpole.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+## Train
+"""
+# Iterate over the number of epochs
+for epoch in range(epochs):
+    # Initialize the sum of the returns, lengths and number of episodes for each epoch
+    sum_ret = 0
+    num_ep = 0
+
+    # Iterate over the steps of each epoch
+    for t in range(steps_per_epoch):
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:190]: Move as much inner code as possible to one or several standalone functions, then decorate them with `@tf.function` compilation decorator. You may also try `@tf.function(jit_compile=True)` to see if it speeds things up.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This will be a great contribution.

File: examples/rl/ppo_cartpole.py (hunk 6 above)
Comment (iliachry) [reply examples/rl/ppo_cartpole.py:219]: Hello Francois, thanks for the useful comments. I changed the variable names and I also moved the code for updating the policy and the value function on standalone functions using `@tf.function.`  The `jit_compile=True` argument did not speed things up, however.

File: examples/rl/ppo_cartpole.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    # Print mean return and length for each epoch
+    print("\n Epoch: " + str(epoch + 1) + ". Mean Return: " + str(sum_return / num_episodes) + ". Mean Length: " + str(sum_length / num_episodes))
+
+
+"""
+## Visualizations
+
+Before training:
+
+![Imgur](https://i.imgur.com/rKXDoMC.gif)
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:274]: Seems to work well enough... before training?

File: examples/rl/ppo_cartpole.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        if kl > 1.5 * target_kl:
+            # Early Stopping
+            break
+
+    # Update the value function
+    for _ in range(train_value_iterations):
+        train_value_function(observation_buffer, return_buffer)
+
+    # Print mean return and length for each epoch
+    print("\n Epoch: " + str(epoch + 1) + ". Mean Return: " + str(sum_return / num_episodes) + ". Mean Length: " + str(sum_length / num_episodes))
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:266]: You'll need to format the file with `black`.

Also, for print lines like this one, you'll find that it's much more readable to use f-strings, e.g.

f"Epoch: {epoch + 1} ..."

Comment (iliachry) [reply examples/rl/ppo_cartpole.py]: I formatted the file with `black` and used f-string for printing. Thanks for the comments.

File: examples/rl/ppo_cartpole.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    num_episodes = 0
+
+    # Iterate over the steps of each epoch
+    for t in range(steps_per_epoch):
+        if render:
+            env.render()
+
+        # Get the logits, action, and take one step in the environment
+        observation = observation.reshape(1, -1)
+        logits = actor(observation)
```
Comment (fchollet) [inline examples/rl/ppo_cartpole.py:257]: You should switch this to `model.predict()` to benefit from compilation (or otherwise move this to inside a tf.function). If you use `model.predict()` you'll get a numpy array, so the line below should be switched to numpy as well.

File: examples/rl/ppo_cartpole.py (hunk 7 above)
Comment (fchollet) [reply examples/rl/ppo_cartpole.py:315]: Is this expected?

File: examples/rl/ppo_cartpole.py (hunk 9 above)
Comment (iliachry) [reply examples/rl/ppo_cartpole.py]: Done!

File: examples/rl/ppo_cartpole.py (hunk 7 above)
Comment (iliachry) [reply examples/rl/ppo_cartpole.py:322]: It does not work well. The environment resets when the pole deviates more than 15 degrees from the vertical axis. However, if you do not know it, it's not easy to understand it. If you see the other two gifs the movement is much more smooth because the environment reset is not so frequent. So, the discontinuity in the "before training" gif shows that the random agent does not work well.

Comment (fchollet) [reply examples/rl/ppo_cartpole.py:322]: Got it.

**Review by fchollet — APPROVED** [review]
LGTM -- please add the generated files.

**Comment by iliachry** [comment]
Added the generated files.

**Comment by fchollet** [comment]
Thanks for the great contribution! 👍

**Comment by iliachry** [comment]
Thanks a lot for the valuable comments. Looking forward to contributing more to keras.
